{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import textwrap\n",
    "from collections import defaultdict\n",
    "from pandas.api.types import CategoricalDtype # for sorting based on the list\n",
    "\n",
    "from pacman.config import RESULTS_DIR\n",
    "\n",
    "# Load data\n",
    "invalid_studies = [\"THYM\", \"CHOL\", \"DLBC\", \"COADREAD_old\"] # [\"PCPG\", \"UCS\"]\n",
    "root = f\"{RESULTS_DIR}/genomic/dseq_results\"\n",
    "studies = sorted([study for study in os.listdir(root) if os.path.isdir(os.path.join(root, study)) and study not in invalid_studies])\n",
    "df_lists = []\n",
    "lists = []\n",
    "\n",
    "for study in studies:\n",
    "    df = pd.read_csv(f\"{root}/{study}/gseapy.gene_set.prerank.report.csv\")\n",
    "    df_lists.append(df)\n",
    "    lists.append(df[\"Term\"].to_list())\n",
    "\n",
    "# Find common names across all lists\n",
    "common_names = set(lists[0])\n",
    "for lst in lists[1:]:\n",
    "    common_names &= set(lst)\n",
    "\n",
    "# Create a dictionary to store the ranks of common names\n",
    "name_ranks = defaultdict(list)\n",
    "\n",
    "# Populate the dictionary with ranks from each list\n",
    "for i, lst in enumerate(lists):\n",
    "    for rank, name in enumerate(lst):\n",
    "        if name in common_names:\n",
    "            name_ranks[name].append(rank)\n",
    "\n",
    "# Calculate the average rank for each common name\n",
    "average_ranks = {name: np.mean(ranks) for name, ranks in name_ranks.items()}\n",
    "\n",
    "# Sort the names by their average rank\n",
    "sorted_names = sorted(average_ranks, key=average_ranks.get)\n",
    "\n",
    "# # Select the top 20 names\n",
    "# top_20_names = sorted_names[:20]\n",
    "# OR\n",
    "# manually editing all the interesting pathways\n",
    "top_20_names = [\n",
    "    # Mitosis & Cell Cycle\n",
    "    'Reactome_2022__Cell Cycle, Mitotic R-HSA-69278',\n",
    "    'Reactome_2022__Mitotic Prometaphase R-HSA-68877',\n",
    "    'Reactome_2022__Mitotic Metaphase And Anaphase R-HSA-2555396',\n",
    "    # 'MSigDB_Hallmark_2020__Mitotic Spindle',\n",
    "    'Reactome_2022__Resolution Of Sister Chromatid Cohesion R-HSA-2500257',\n",
    "    'Reactome_2022__EML4 And NUDC In Mitotic Spindle Formation R-HSA-9648025',\n",
    "    'Reactome_2022__Mitotic Spindle Checkpoint R-HSA-69618',\n",
    "    'MSigDB_Hallmark_2020__G2-M Checkpoint',\n",
    "    # 'Reactome_2022__Unattached Kinetochores Signal Amplification Via A MAD2 Inhibitory Signal R-HSA-141444', # maybe remove from here?\n",
    "    'MSigDB_Hallmark_2020__E2F Targets',\n",
    "    'MSigDB_Hallmark_2020__Myc Targets V1',\n",
    "\n",
    "    # DNA Replication & Repair\n",
    "    'Reactome_2022__Activation Of ATR In Response To Replication Stress R-HSA-176187',\n",
    "    'Reactome_2022__DNA Strand Elongation R-HSA-69190',\n",
    "    'Reactome_2022__DNA Replication R-HSA-69306',\n",
    "    'Reactome_2022__Homology Directed Repair R-HSA-5693538',\n",
    "\n",
    "    # Gene Expression and Protein Synthesis\n",
    "    'Reactome_2022__Processing Of Capped Intron-Containing Pre-mRNA R-HSA-72203',\n",
    "    'Reactome_2022__mRNA Splicing R-HSA-72172',\n",
    "    'Reactome_2022__FGFR2 Alternative Splicing R-HSA-6803529',\n",
    "    'Reactome_2022__Formation Of A Pool Of Free 40S Subunits R-HSA-72689',\n",
    "    'Reactome_2022__GTP Hydrolysis And Joining Of 60S Ribosomal Subunit R-HSA-72706',\n",
    "    'Reactome_2022__Translation R-HSA-72766',\n",
    "    'Reactome_2022__Peptide Chain Elongation R-HSA-156902',\n",
    "    'Reactome_2022__Eukaryotic Translation Termination R-HSA-72764',\n",
    "    'Reactome_2022__Nonsense Mediated Decay (NMD) Independent Of Exon Junction Complex (EJC) R-HSA-975956',\n",
    "\n",
    "    # Metabolism and Energy\n",
    "    'MSigDB_Hallmark_2020__mTORC1 Signaling',\n",
    "    # 'Reactome_2022__Metabolic Disorders Of Biological Oxidation Enzymes R-HSA-5579029',\n",
    "    # 'Reactome_2022__Metabolism Of Steroid Hormones R-HSA-196071',\n",
    "    'Reactome_2022__Glycolysis R-HSA-70171',\n",
    "    'MSigDB_Hallmark_2020__Oxidative Phosphorylation',\n",
    "    'Reactome_2022__Respiratory Electron Transport R-HSA-611105',\n",
    "    'Reactome_2022__Selenoamino Acid Metabolism R-HSA-2408522',\n",
    "    \n",
    "\n",
    "    # Cancer Progression and Metastasis\n",
    "    'MSigDB_Hallmark_2020__Epithelial Mesenchymal Transition',\n",
    "\n",
    "    # Immune Response and Inflammation\n",
    "    'MSigDB_Hallmark_2020__Interferon Gamma Response',\n",
    "    'MSigDB_Hallmark_2020__Inflammatory Response',\n",
    "    'MSigDB_Hallmark_2020__TNF-alpha Signaling via NF-kB',\n",
    "]\n",
    "# Define the categorical type with the desired order\n",
    "cat_type = CategoricalDtype(categories=top_20_names, ordered=True)\n",
    "print(\"Top 20 common names:\", top_20_names)\n",
    "\n",
    "# Function to clean term names\n",
    "def clean_term(term):\n",
    "    parts = term.split(\"__\")\n",
    "    main_term = parts[1] if len(parts) > 1 else term\n",
    "    if parts[0] == \"Reactome_2022\" and \"R-HSA-\" in main_term:\n",
    "        main_term = main_term.split(\"R-HSA-\")[0].strip()\n",
    "    return \"\\n\".join(textwrap.wrap(main_term, width=42))\n",
    "\n",
    "# Prepare data for plotting\n",
    "plot_data = []\n",
    "\n",
    "for study, df in zip(studies, df_lists):\n",
    "    df_filtered = df[df[\"Term\"].isin(top_20_names)].copy()\n",
    "\n",
    "    # Convert 'Term' to the categorical type and Sort the DataFrame based on 'Term'\n",
    "    df_filtered['Term'] = df_filtered['Term'].astype(cat_type)\n",
    "    df_filtered = df_filtered.sort_values('Term')\n",
    "\n",
    "    df_filtered[\"-log(FDR)\"] = -np.log10(df_filtered[\"FDR q-val\"].replace(0, 1e-10))\n",
    "    df_filtered[\"Study\"] = study\n",
    "    df_filtered[\"Clean_Term\"] = df_filtered[\"Term\"].apply(clean_term)\n",
    "    # df_filtered = df_filtered.sort_values(by=\"Clean_Term\")\n",
    "    plot_data.append(df_filtered)\n",
    "\n",
    "plot_data = pd.concat(plot_data)\n",
    "# plot_data = plot_data.sort_values(by=\"Clean_Term\")\n",
    "\n",
    "# Plot the dotplot\n",
    "# plt.figure(figsize=(10, 6))\n",
    "fig, ax =plt.subplots(1,1,figsize=(7, 8.5))\n",
    "dotplot = sns.scatterplot(\n",
    "    data=plot_data,\n",
    "    x=\"Study\",\n",
    "    y=\"Clean_Term\",\n",
    "    size=\"-log(FDR)\",\n",
    "    hue=\"NES\",\n",
    "    palette=\"PiYG\",\n",
    "    sizes=(20, 200),\n",
    "    edgecolor=\"gray\",\n",
    "    linewidth=0.5,\n",
    "    ax=ax\n",
    ")\n",
    "# Adjust line spacing for y-axis tick labels\n",
    "for label in ax.get_yticklabels():\n",
    "    label.set_linespacing(0.8)  # Reduce line spacing\n",
    "    \n",
    "ax.margins(y=0.02, x=0.03)\n",
    "dotplot.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)\n",
    "plt.xticks(rotation=90)\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"\")\n",
    "# plt.title('Dotplot of Top Common Terms across Studies')\n",
    "plt.grid(True, which='both', linestyle='--', linewidth=0.5)\n",
    "plt.gca().set_axisbelow(True)\n",
    "\n",
    "# plt.tight_layout()\n",
    "\n",
    "plt.savefig(f\"{root}/dseq_results_complete_wide.png\", dpi=600, bbox_inches = 'tight', pad_inches = 0.01)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
